### DISCUSSION TITLE
Does iceoryx support Android NDK?

### DISCUSSION THREAD

**Discussion by mertonmeng** [eclipse-iceoryx/iceoryx#1901] [Q&A]
Hi there,

I am considering to use iceoryx as IPC library for a cross-platform project. It will need to support Android, Arm Linux 64 and X86 Linux. Arm Linux 64 and X86 Linux seems already supported. I am wondering if iceoryx support Android NDK?

Thanks!

**Comment by mossmaurice** [comment]
Hi @mertonmeng 

Thanks for starting this discussion! 

Android support has been requested before, see [here](https://github.com/eclipse-iceoryx/iceoryx/issues/942). However, there was no final decision if this feature will be supported, also AFAIK no one started to work on it yet. The iceoryx maintainers will not work on it in the near future.

Of course, your contribution of this feature would be much appreciated! :blush: You can find more details on a port for a new platform [over here](https://github.com/eclipse-iceoryx/iceoryx/discussions/1381#discussioncomment-2878833).

If you plan to work on this feature it would probably best to discuss it beforehand in the [developer meetup](https://github.com/eclipse-iceoryx/iceoryx/blob/master/GOVERNANCE.md). The next one is today at 17:00 CET.

**Reply by elfenpiff** [reply]
@mertonmeng @mossmaurice This is not entirely true. I started exploring android in the beginning of last year an it seems that the only thing which is not working right out of the box is the shared memory.
Maybe it is easier than we think when this is the only issue. In the end it is just a linux system.

@mertonmeng What kind of project do you have in mind you need iceoryx for?

**Reply by mertonmeng** [reply]
Thanks for the quick reply @elfenpiff @mossmaurice !
Let me deep dive a little on how I can port them.
Due to confidentiality of the project, I cannot disclose any details. Sorry about that.

**Reply by elfenpiff** [reply]
@mertonmeng If you like you could start creating a new platform by following this article: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/custom-iceoryx-platform.md

I think it may can be done quickly. Just copy the linux platform folder to a folder called android and try to compile everything. I assume that `shm_***` inside `iox_shm_**` will maybe cause some issues. If this is the case you can remove the implementation (forwarding) in the platform so that you have just an empty function. If everything compiles you can forward the shared memory function to the android pendant.

If this is done, execute the `iceoryx_hoofs` unit tests. If they pass then iceoryx should work, if something fails you have to look at the platform again and fix the posix construct which is causing the problem.

And of course if you need support with any of the steps please do not hesitate to ask  us here.

**Reply by mertonmeng** [reply]
Thanks a lot for the instructions! @elfenpiff 
I think maybe shm_open can be replaced by the ashmem_create in here: https://elinux.org/Android_Kernel_Features
But not sure about shm_unlink
I will follow your steps first and see if there is anything missing.

**Comment by mertonmeng** [comment]
Hi @elfenpiff @mossmaurice 
I have tried your suggestion, basically just copy the linux posix functions to android, and removed shm_*** to replace with the dummy value. So I am setting up cmake with NDK and all:
`cmake -Bbuild -Hiceoryx_meta -DIOX_PLATFORM_PATH=/home/iceoryx/iceoryx/iceoryx_platform/android \
-DCMAKE_TOOLCHAIN_FILE=/opt/android-ndk-r21b/build/cmake/android.toolchain.cmake \
-DANDROID_ABI=arm64-v8a \
-DANDROID_NATIVE_API_LEVEL=24`
But got the error:

> -- Detecting C compiler ABI info
> -- Detecting C compiler ABI info - done
> -- Check for working C compiler: /opt/android-ndk-r21b/toolchains/llvm/prebuilt/linux-x86_64/bin/clang - skipped
> -- Detecting C compile features
> -- Detecting C compile features - done
> -- Detecting CXX compiler ABI info
> -- Detecting CXX compiler ABI info - done
> -- Check for working CXX compiler: /opt/android-ndk-r21b/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++ - skipped
> -- Detecting CXX compile features
> -- Detecting CXX compile features - done
> -- The package 'iceoryx_platform' is used in source code version.
> -- [i] <<<<<<<<<<<<< Start iceoryx_hoofs configuration: >>>>>>>>>>>>>
> -- [i] IOX_MINIMAL_LOG_LEVEL: TRACE
> -- [i] <<<<<<<<<<<<<< End iceoryx_hoofs configuration: >>>>>>>>>>>>>>
> -- The package 'iceoryx_hoofs' is used in source code version.
> -- Performing Test CMAKE_HAVE_LIBC_PTHREAD
> -- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
> -- Looking for pthread_create in pthreads
> -- Looking for pthread_create in pthreads - not found
> -- Looking for pthread_create in pthread
> -- Looking for pthread_create in pthread - not found
> -- Check if compiler accepts -pthread
> -- Check if compiler accepts -pthread - yes
> -- Found Threads: TRUE  
> -- The package 'iceoryx_dust' is used in source code version.
> -- Configuring done
> -- Generating done
> -- Build files have been written to: /home/iceoryx/iceoryx/build/dependencies/cpptoml/download
> [ 11%] Creating directories for 'ext_cpptoml'
> [ 22%] Performing download step (git clone) for 'ext_cpptoml'
> Cloning into 'src'...
> HEAD is now at fededad one-char typo
> Submodule 'deps/meta-cmake' (https://github.com/meta-toolkit/meta-cmake.git) registered for path 'deps/meta-cmake'
> Cloning into '/home/iceoryx/iceoryx/build/dependencies/cpptoml/src/deps/meta-cmake'...
> Submodule path 'deps/meta-cmake': checked out 'db7ff976c430a6cb6cbd03ae691be04cfaac0a07'
> [ 33%] Performing update step for 'ext_cpptoml'
> [ 44%] No patch step for 'ext_cpptoml'
> [ 55%] No configure step for 'ext_cpptoml'
> [ 66%] No build step for 'ext_cpptoml'
> [ 77%] No install step for 'ext_cpptoml'
> [ 88%] No test step for 'ext_cpptoml'
> [100%] Completed 'ext_cpptoml'
> [100%] Built target ext_cpptoml
> -- ANDROID_PLATFORM not set. Defaulting to minimum supported version
> 16.
> -- Detecting C compiler ABI info
> -- Detecting C compiler ABI info - done
> -- Check for working C compiler: /opt/android-ndk-r21b/toolchains/llvm/prebuilt/linux-x86_64/bin/clang - skipped
> -- Detecting C compile features
> -- Detecting C compile features - done
> -- Detecting CXX compiler ABI info
> -- Detecting CXX compiler ABI info - done
> -- Check for working CXX compiler: /opt/android-ndk-r21b/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++ - skipped
> -- Detecting CXX compile features
> -- Detecting CXX compile features - done
> -- Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE) 
> -- Configuring done
> -- Generating done
> -- Build files have been written to: /home/iceoryx/iceoryx/build/dependencies/cpptoml/build
> Install the project...
> -- Install configuration: ""
> -- Installing: /home/iceoryx/iceoryx/build/dependencies/install/include/cpptoml.h
> -- Installing: /home/iceoryx/iceoryx/build/dependencies/install/lib/cmake/cpptoml/cpptomlTargets.cmake
> -- Installing: /home/iceoryx/iceoryx/build/dependencies/install/lib/cmake/cpptoml/cpptomlConfigVersion.cmake
> -- Installing: /home/iceoryx/iceoryx/build/dependencies/install/lib/cmake/cpptoml/cpptomlConfig.cmake
> CMake Error at /home/iceoryx/iceoryx/iceoryx_posh/CMakeLists.txt:37 (find_package):
>   Could not find a configuration file for package "cpptoml" that is
>   compatible with requested version "".
>   The following configuration files were considered but not accepted:
>     /home/iceoryx/iceoryx/build/dependencies/install/lib/cmake/cpptoml/cpptomlConfig.cmake, version: 0.4.0 (32bit)
> -- Configuring incomplete, errors occurred!
> See also "/home/iceoryx/iceoryx/build/CMakeFiles/CMakeOutput.log".
> See also "/home/iceoryx/iceoryx/build/CMakeFiles/CMakeError.log".

Any suggestions?

**Reply by elfenpiff** [reply]
@mertonmeng You can disable tom file config support by calling cmake with the additional switch `-DTOML_CONFIG=OFF`. 

Do you have a branch where you are working on? If something else comes up you could just provide the link to it here to make things easier.

**Comment by uupks** [comment]
It can't access shared memory by file name with Ashmem on Android platform.

I have cross compiled iceoryx and run iox-roudi successfully on Android, but failed to map created shared memory into process when running icehello publisher example.

```
2023-04-13 19:43:18.010 [Warn ]: IPC channel still there, doing an unlink of iox-cpp-publisher-helloworld
fd : 6
2023-04-13 19:43:18.011 [Error]: /home/r10051/01_code/iceoryx/iceoryx_hoofs/source/posix_wrapper/shared_memory_object/memory_map.cpp:33 { expected<iox::posix::MemoryMap, iox::posix::MemoryMapError> iox::posix::MemoryMapBuilder::create() -> mmap }  :::  [ 22 ]  Invalid argument
2023-04-13 19:43:18.011 [Error]: Unable to map memory with the following properties [ baseAddressHint = 0x0, length = 66611176, fileDescriptor = 6, access mode = AccessMode::READ_WRITE, flags = 00000000000000000000000000000001, offset = 0x0 ]
2023-04-13 19:43:18.011 [Error]: One or more of the following failures happened:
  1. The address, length or the offset is not aligned on a page boundary.
  2. The provided length is 0.
  3. One of the flags of MAP_PRIVATE, MAP_SHARED or MAP_SHARED_VALIDATE is missing.
2023-04-13 19:43:18.011 [Error]: Unable to create a shared memory object with the following properties [ name = iceoryx_mgmt, sizeInBytes = 66611176, access mode = AccessMode::READ_WRITE, open mode = OpenMode::OPEN_EXISTING, baseAddressHint =  (no hint set), permissions = 0o660 ]
2023-04-13 19:43:18.011 [Error]: Failed to map created shared memory into process!
2023-04-13 19:43:18.011 [Error]: ICEORYX error! POSH__SHM_APP_MAPP_ERR
terminating
Aborted 
```

**Reply by elfenpiff** [reply]
@uupks did you also create a new android platform, ported the `mmap.cpp|hpp` code and used this when you cross-compiled iceoryx? I am asking since android seems to support only the system V shared memory API and iceoryx natively uses the posix API, so unless you ported the the `mmap.cpp|hpp` code inside the platform, it will not work.

Are you currently working on such a platform and require some assistance?

**Reply by uupks** [reply]
yes, I followed your instructions and ported mmap.cpp and mmap.hpp.

I implemented `iox_shm_open` with `ashmem_create_region` , and `iox_shm_unlink` with `return 0`.

**Reply by elfenpiff** [reply]
@uupks do you have a public branch where I can take a look at the implementation?

**Reply by elfenpiff** [reply]
@uupks Before trying to run the examples could you please execute the iceoryx_hoofs unit tests and see if everything is compliant. If a test there is failing most examples will probably fail to work.
To run the test, do the following:
```sh
cmake -Bbuild -Hiceoryx_meta -DBUILD_TEST=ON
cd build
make
./hoofs/test/hoofs_moduletests
```

I assume several tests will still fail but they will provide us some insights what needs to be adjusted for android. Could you maybe add here the tests which still fail.

**Reply by elfenpiff** [reply]
@uupks It seems that `ashmem_create_region` requires a size argument but `iox_shm_open` does not has one. So a one to one mapping will not work, we have to do some deeper adjustments in the code here.

**Reply by uupks** [reply]
https://github.com/uupks/iceoryx/tree/feature/android_support

**Reply by uupks** [reply]
I use colcon and android ndk 23c to build the project, so you need to install these firstly.

```
export NDK_HOME=
./build.sh
```

**Reply by uupks** [reply]
> @uupks It seems that `ashmem_create_region` requires a size argument but `iox_shm_open` does not has one. So a one to one mapping will not work, we have to do some deeper adjustments in the code here.

yes, it can also use ioctl to change the memory size

```
int ftruncate(int fildes, off_t length)
{
    return ioctl(fildes, ASHMEM_SET_SIZE, length);
}
```

**Reply by uupks** [reply]
according to this https://github.com/SHallimani/Android-NDK-Examples#answer

> Ashmem doesn't work like regular shared memory on Linux, and there is a good reason for it. First, let's try to explain the "(deleted)" part, this is an implementation detail of how ashmem is implemented in the kernel. What it really means is that a file entry was created in the /dev/ashmem/ directory, then later removed, but that the corresponding i-node still exists because there is at least one open file-descriptor for it. You could actually create several ashmem regions with the same name, and they would all appear as "/dev/ashmem/ (deleted)", but each one of them would correspond to a different i-node, and thus a different memory region. And if you look under /dev/ashmem/ you would see that the directory is still empty. That's why the name of an ashmem region is really only used for debugging. There is no way to 'open' an existing region by name. 

It seems that iceoryx apps access shared memory by opening the same name "iceoryx_mgmt" that is created by iox-roudi.

This might not work on Android with ashmem.

**Reply by elBoberido** [reply]
@uupks I'm not quite sure whether this is an issue. Are the files in `/dev/ashmem` automatically deleted even when the process who created them is still running? If not, then it should not be a problem since they are only explicitly deleted by RouDi when it shuts down gracefully.

**Reply by uupks** [reply]
The /dev/ashmem/ directory is still empty after running iox-roudi...

**Reply by elBoberido** [reply]
Did you manually try to create an entry in `/dev/ashmem/`? Maybe the mechanism in RouDi is broken on Android or not yet correctly implemented. Having a shared memory which cannot be accessed by other processes does not really make sense.

**Reply by elBoberido** [reply]
Did you check this sources?
https://stackoverflow.com/questions/53058077/multiprocess-by-using-android-ndk-sharedmemory-h/53062250#53062250
https://devarea.com/android-creating-shared-memory-using-ashmem/

**Reply by uupks** [reply]
you can check my public branch mentioned above.


```
int iox_shm_open(const char* name, int oflag, mode_t mode)
{
    int fd, ret;
    fd = open(ASHMEM_DEVICE, O_RDWR);
    if (fd < 0) {
        return fd;
    }

    if (name) {
        char buf[ASHMEM_NAME_LEN] = {0};

        strlcpy(buf, name, sizeof(buf));
        ret = ioctl(fd, ASHMEM_SET_NAME, buf);
        if (ret < 0) {
            goto error;
        }
    }

    return fd;

error:
    close(fd);
    return ret;
}
```

use ioctl to control the shared memory size.
```
int ftruncate(int fildes, off_t length)
{
    return ioctl(fildes, ASHMEM_SET_SIZE, length);
}
```

I think the problem is it can only access shared memory by transport the fd to another process on Android system.

We can't access the shared memory region by opening the same device name like `/dev/ashmem/name`

**Reply by elBoberido** [reply]
The first link I posted suggested to use `ASharedMemory_create`. Did you try that?

I have no experience with Android, so it might indeed not work and exchanging FDs via binder might be necessary. If this is the case, it still might be possible to do this behind the scene in the platform abstraction but it will definitely become ugly.

**Reply by uupks** [reply]
`ASharedMemory_create` is actually a wrapper of what i have done ...

https://github.com/termux/libandroid-shmem

https://github.com/green-green-avk/libwrapdroid

There two repos use unix domain socket to emulate SystemV and POSIX shared memory on Android using ashmem.

Can we also do this in iox-roudi?

**Reply by elBoberido** [reply]
It might be possible but I'm not sure how ugly it will become to do all of this in the platform layer. I guess it is worth a try.
